File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/timer_recordings/TimerRecordingAddEditFragment.java
package org.tvheadend.tvhclient.ui.features.dvr.timer_recordings;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.Channel;
import org.tvheadend.tvhclient.domain.entity.ServerProfile;
import org.tvheadend.tvhclient.domain.entity.TimerRecording;
import org.tvheadend.tvhclient.data.service.HtspService;
import org.tvheadend.tvhclient.ui.features.dvr.DatePickerFragment;
import org.tvheadend.tvhclient.ui.features.dvr.RecordingConfigSelectedListener;
import org.tvheadend.tvhclient.ui.features.dvr.RecordingUtils;
import org.tvheadend.tvhclient.ui.features.dvr.TimePickerFragment;
import org.tvheadend.tvhclient.ui.base.BaseFragment;
import org.tvheadend.tvhclient.ui.base.callbacks.BackPressedInterface;
import org.tvheadend.tvhclient.util.MiscUtils;
import org.tvheadend.tvhclient.ui.base.utils.SnackbarUtils;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProviders;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnTextChanged;
import butterknife.Unbinder;

public class TimerRecordingAddEditFragment extends BaseFragment implements BackPressedInterface, RecordingConfigSelectedListener, DatePickerFragment.Listener, TimePickerFragment.Listener {

    @BindView(R.id.is_enabled)
    CheckBox isEnabledCheckbox;
    @BindView(R.id.priority)
    TextView priorityTextView;
    @BindView(R.id.days_of_week)
    TextView daysOfWeekTextView;
    @BindView(R.id.start_time)
    TextView startTimeTextView;
    @BindView(R.id.stop_time)
    TextView stopTimeTextView;
    @BindView(R.id.directory)
    EditText directoryEditText;
    @BindView(R.id.directory_label)
    TextView directoryLabelTextView;
    @BindView(R.id.title)
    EditText titleEditText;
    @BindView(R.id.name)
    EditText nameEditText;
    @BindView(R.id.channel)
    TextView channelNameTextView;
    @BindView(R.id.dvr_config)
    TextView recordingProfileNameTextView;
    @BindView(R.id.dvr_config_label)
    TextView recordingProfileLabelTextView;

    private String[] recordingProfilesList;
    private List<Channel> channelList;
    private ServerProfile profile;
    private int recordingProfileNameId;

    private Unbinder unbinder;
    private String id;
    private TimerRecording recording;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.timer_recording_add_edit_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        recordingProfilesList = appRepository.getServerProfileData().getRecordingProfileNames();
        profile = appRepository.getServerProfileData().getItemById(serverStatus.getRecordingServerProfileId());
        recordingProfileNameId = RecordingUtils.getSelectedProfileId(profile, recordingProfilesList);
        channelList = appRepository.getChannelData().getItems();

        if (savedInstanceState != null) {
            id = savedInstanceState.getString("id", "");
            recordingProfileNameId = savedInstanceState.getInt("configName");
        } else {
            Bundle bundle = getArguments();
            if (bundle != null) {
                id = bundle.getString("id", "");
            }
        }

        setHasOptionsMenu(true);

        // Get the recoding from the view model without observing it.
        // This allows storing any changes and preserving them during orientation changes.
        TimerRecordingViewModel viewModel = ViewModelProviders.of(activity).get(TimerRecordingViewModel.class);
        recording = viewModel.getRecordingByIdSync(id);
        updateUI();

        toolbarInterface.setTitle(!TextUtils.isEmpty(id) ?
                getString(R.string.edit_recording) :
                getString(R.string.add_recording));
    }

    private void updateUI() {
        int gmtOffset = serverStatus.getGmtoffset();

        isEnabledCheckbox.setVisibility(htspVersion >= 19 ? View.VISIBLE : View.GONE);
        isEnabledCheckbox.setChecked(recording.isEnabled());
        titleEditText.setText(recording.getTitle());
        nameEditText.setText(recording.getName());

        directoryLabelTextView.setVisibility(htspVersion >= 19 ? View.VISIBLE : View.GONE);
        directoryEditText.setVisibility(htspVersion >= 19 ? View.VISIBLE : View.GONE);
        directoryEditText.setText(recording.getDirectory());

        channelNameTextView.setText(!TextUtils.isEmpty(recording.getChannelName()) ? recording.getChannelName() : getString(R.string.all_channels));
        channelNameTextView.setOnClickListener(view -> {
            // Determine if the server supports recording on all channels
            boolean allowRecordingOnAllChannels = htspVersion >= 21;
            RecordingUtils.handleChannelListSelection(activity, channelList, allowRecordingOnAllChannels, TimerRecordingAddEditFragment.this);
        });

        priorityTextView.setText(RecordingUtils.getPriorityName(activity, recording.getPriority()));
        priorityTextView.setOnClickListener(view -> RecordingUtils.handlePrioritySelection(activity, recording.getPriority(), TimerRecordingAddEditFragment.this));

        if (recordingProfilesList.length == 0) {
            recordingProfileNameTextView.setVisibility(View.GONE);
            recordingProfileLabelTextView.setVisibility(View.GONE);
        } else {
            recordingProfileNameTextView.setVisibility(View.VISIBLE);
            recordingProfileLabelTextView.setVisibility(View.VISIBLE);

            recordingProfileNameTextView.setText(recordingProfilesList[recordingProfileNameId]);
            recordingProfileNameTextView.setOnClickListener(view -> RecordingUtils.handleRecordingProfileSelection(activity, recordingProfilesList, recordingProfileNameId, TimerRecordingAddEditFragment.this));
        }

        startTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(recording.getStart() - gmtOffset));
        startTimeTextView.setOnClickListener(view -> RecordingUtils.handleTimeSelection(activity, recording.getStart() - gmtOffset, TimerRecordingAddEditFragment.this, "startTime"));

        stopTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(recording.getStop() - gmtOffset));
        stopTimeTextView.setOnClickListener(view -> RecordingUtils.handleTimeSelection(activity, recording.getStop() - gmtOffset, TimerRecordingAddEditFragment.this, "stopTime"));

        daysOfWeekTextView.setText(RecordingUtils.getSelectedDaysOfWeekText(activity, recording.getDaysOfWeek()));
        daysOfWeekTextView.setOnClickListener(view -> RecordingUtils.handleDayOfWeekSelection(activity, recording.getDaysOfWeek(), TimerRecordingAddEditFragment.this));
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        outState.putString("id", id);
        outState.putInt("configName", recordingProfileNameId);
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.save_cancel_options_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                cancel();
                return true;
            case R.id.menu_save:
                save();
                return true;
            case R.id.menu_cancel:
                cancel();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void save() {
        if (TextUtils.isEmpty(recording.getTitle())) {
            SnackbarUtils.sendSnackbarMessage(activity, R.string.error_empty_title);
            return;
        }
        if (!TextUtils.isEmpty(id)) {
            updateTimerRecording();
        } else {
            addTimerRecording();
        }
    }

    private void cancel() {
        // Show confirmation dialog to cancel
        new MaterialDialog.Builder(activity)
                .content(R.string.cancel_add_recording)
                .positiveText(getString(R.string.discard))
                .negativeText(getString(R.string.cancel))
                .onPositive((dialog, which) -> activity.finish())
                .onNegative((dialog, which) -> dialog.cancel())
                .show();
    }

    /**
     * Adds a new timer recording with the given values. This method is also
     * called when a recording is being edited. It adds a recording with edited
     * values which was previously removed.
     */
    private void addTimerRecording() {
        Intent intent = getIntentData();
        intent.setAction("addTimerecEntry");
        activity.startService(intent);
        activity.finish();
    }

    /**
     * Updates the timer recording with the given values.
     * If the API version supports it, use the native service call method
     * otherwise the old recording is removed and a new one with the
     * edited values is added afterwards. This is done in the service
     */
    private void updateTimerRecording() {
        Intent intent = getIntentData();
        intent.setAction("updateTimerecEntry");
        intent.putExtra("id", id);
        activity.startService(intent);
        activity.finish();
    }

    /**
     * Returns an intent with the recording data
     */
    private Intent getIntentData() {
        Intent intent = new Intent(activity, HtspService.class);
        intent.putExtra("directory", recording.getDirectory());
        intent.putExtra("title", recording.getTitle());
        intent.putExtra("name", recording.getName());
        intent.putExtra("start", RecordingUtils.getMinutesFromTimeInMillis(recording.getStart()));
        intent.putExtra("stop", RecordingUtils.getMinutesFromTimeInMillis(recording.getStop()));
        intent.putExtra("daysOfWeek", recording.getDaysOfWeek());
        intent.putExtra("priority", recording.getPriority());
        intent.putExtra("enabled", recording.isEnabled() ? 1 : 0);

        if (recording.getChannelId() > 0) {
            intent.putExtra("channelId", recording.getChannelId());
        }

        // Add the recording profile if available and enabled
        if (MiscUtils.isServerProfileEnabled(profile, serverStatus)
                && (recordingProfileNameTextView.getText().length() > 0)) {
            // Use the selected profile. If no change was done in the 
            // selection then the default one from the connection setting will be used
            intent.putExtra("configName", recordingProfileNameTextView.getText().toString());
        }
        return intent;
    }

    @Override
    public void onChannelSelected(Channel channel) {
        recording.setChannelId(channel.getId());
        channelNameTextView.setText(channel.getName());
    }

    @Override
    public void onPrioritySelected(int which) {
        recording.setPriority(which);
        priorityTextView.setText(RecordingUtils.getPriorityName(activity, recording.getPriority()));
    }

    @Override
    public void onProfileSelected(int which) {
        recordingProfileNameTextView.setText(recordingProfilesList[which]);
        recordingProfileNameId = which;
    }

    @Override
    public void onTimeSelected(long milliSeconds, String tag) {
        if (tag.equals("startTime")) {
            recording.setStart(milliSeconds);
            startTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(milliSeconds));

            // If the start time is after the stop time,
            // update the stop time with the start value
            if (milliSeconds > recording.getStop()) {
                recording.setStop(milliSeconds);
                stopTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(milliSeconds));
            }

        } else if (tag.equals("stopTime")) {
            recording.setStop(milliSeconds);
            stopTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(milliSeconds));

            // If the stop time is before the start time,
            // update the start time with the stop value
            if (milliSeconds < recording.getStart()) {
                recording.setStart(milliSeconds);
                startTimeTextView.setText(RecordingUtils.getTimeStringFromTimeInMillis(milliSeconds));
            }
        }
    }

    @Override
    public void onDateSelected(long milliSeconds, String tag) {
        // NOP
    }

    @Override
    public void onDaysSelected(int selectedDays) {
        recording.setDaysOfWeek(selectedDays);
        daysOfWeekTextView.setText(RecordingUtils.getSelectedDaysOfWeekText(activity, selectedDays));
    }

    @Override
    public void onBackPressed() {
        cancel();
    }

    @OnTextChanged(R.id.title)
    void onTitleTextChanged(CharSequence text, int start, int count, int after) {
        recording.setTitle(text.toString());
    }

    @OnTextChanged(R.id.name)
    void onNameTextChanged(CharSequence text, int start, int count, int after) {
        recording.setName(text.toString());
    }

    @OnTextChanged(R.id.directory)
    void onDirectoryTextChanged(CharSequence text, int start, int count, int after) {
        recording.setDirectory(text.toString());
    }

    @OnCheckedChanged(R.id.is_enabled)
    void onEnabledCheckboxChanged(CompoundButton buttonView, boolean isChecked) {
        recording.setEnabled(isChecked);
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/dvr/timer_recordings/TimerRecordingAddEditFragment.kt
package org.tvheadend.tvhclient.ui.features.dvr.timer_recordings

import android.os.Bundle
import android.view.*
import androidx.core.view.forEach
import androidx.lifecycle.ViewModelProvider
import com.afollestad.materialdialogs.MaterialDialog
import org.tvheadend.data.entity.Channel
import org.tvheadend.data.entity.ServerProfile
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.databinding.TimerRecordingAddEditFragmentBinding
import org.tvheadend.tvhclient.ui.base.BaseFragment
import org.tvheadend.tvhclient.ui.common.interfaces.BackPressedInterface
import org.tvheadend.tvhclient.ui.common.interfaces.HideNavigationDrawerInterface
import org.tvheadend.tvhclient.ui.common.interfaces.LayoutControlInterface
import org.tvheadend.tvhclient.ui.features.dvr.*
import org.tvheadend.tvhclient.util.extensions.afterTextChanged
import org.tvheadend.tvhclient.util.extensions.sendSnackbarMessage
import org.tvheadend.tvhclient.util.extensions.visibleOrGone
import timber.log.Timber

class TimerRecordingAddEditFragment : BaseFragment(), BackPressedInterface, RecordingConfigSelectedListener, DatePickerFragment.Listener, TimePickerFragment.Listener, HideNavigationDrawerInterface {

    private lateinit var binding: TimerRecordingAddEditFragmentBinding
    private lateinit var timerRecordingViewModel: TimerRecordingViewModel
    private lateinit var recordingProfilesList: Array<String>

    private var profile: ServerProfile? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = TimerRecordingAddEditFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        timerRecordingViewModel = ViewModelProvider(requireActivity()).get(TimerRecordingViewModel::class.java)

        if (activity is LayoutControlInterface) {
           (activity as LayoutControlInterface).forceSingleScreenLayout()
        }

        recordingProfilesList = timerRecordingViewModel.getRecordingProfileNames()
        profile = timerRecordingViewModel.getRecordingProfile()
        timerRecordingViewModel.recordingProfileNameId = getSelectedProfileId(profile, recordingProfilesList)

        if (savedInstanceState == null) {
            timerRecordingViewModel.loadRecordingByIdSync(arguments?.getString("id", "") ?: "")
        }

        updateUI()

        toolbarInterface.setSubtitle("")
        toolbarInterface.setTitle(if (timerRecordingViewModel.recording.id.isNotEmpty())
            getString(R.string.edit_recording)
        else
            getString(R.string.add_recording))
    }

    private fun updateUI() {
        val ctx = context ?: return

        binding.isEnabled.visibleOrGone(htspVersion >= 19)
        binding.isEnabled.isChecked = timerRecordingViewModel.recording.isEnabled

        binding.title.setText(timerRecordingViewModel.recording.title)
        binding.name.setText(timerRecordingViewModel.recording.name)

        binding.directoryLabel.visibleOrGone(htspVersion >= 19)
        binding.directory.visibleOrGone(htspVersion >= 19)
        binding.directory.setText(timerRecordingViewModel.recording.directory)

        binding.channelName.text = timerRecordingViewModel.recording.channelName ?: getString(R.string.all_channels)
        binding.channelName.setOnClickListener {
            // Determine if the server supports recording on all channels
            val allowRecordingOnAllChannels = htspVersion >= 21
            handleChannelListSelection(ctx, timerRecordingViewModel.getChannelList(), allowRecordingOnAllChannels, this@TimerRecordingAddEditFragment)
        }

        binding.priority.text = getPriorityName(ctx, timerRecordingViewModel.recording.priority)
        binding.priority.setOnClickListener {
            handlePrioritySelection(ctx, timerRecordingViewModel.recording.priority, this@TimerRecordingAddEditFragment)
        }

        binding.dvrConfig.visibleOrGone(recordingProfilesList.isNotEmpty())
        binding.dvrConfigLabel.visibleOrGone(recordingProfilesList.isNotEmpty())

        if (recordingProfilesList.isNotEmpty()) {
            binding.dvrConfig.text = recordingProfilesList[timerRecordingViewModel.recordingProfileNameId]
            binding.dvrConfig.setOnClickListener {
                handleRecordingProfileSelection(ctx, recordingProfilesList, timerRecordingViewModel.recordingProfileNameId, this@TimerRecordingAddEditFragment)
            }
        }

        binding.startTime.text = getTimeStringFromTimeInMillis(timerRecordingViewModel.startTimeInMillis)
        binding.startTime.setOnClickListener {
            handleTimeSelection(activity, timerRecordingViewModel.startTimeInMillis, this@TimerRecordingAddEditFragment, "startTime")
        }

        binding.stopTime.text = getTimeStringFromTimeInMillis(timerRecordingViewModel.stopTimeInMillis)
        binding.stopTime.setOnClickListener {
            handleTimeSelection(activity, timerRecordingViewModel.stopTimeInMillis, this@TimerRecordingAddEditFragment, "stopTime")
        }

        binding.daysOfWeek.text = getSelectedDaysOfWeekText(ctx, timerRecordingViewModel.recording.daysOfWeek)
        binding.daysOfWeek.setOnClickListener {
            handleDayOfWeekSelection(ctx, timerRecordingViewModel.recording.daysOfWeek, this@TimerRecordingAddEditFragment)
        }

        binding.timeEnabled.isChecked = timerRecordingViewModel.isTimeEnabled
        handleTimeEnabledClick(binding.timeEnabled.isChecked)

        binding.timeEnabled.setOnClickListener {
            handleTimeEnabledClick(binding.timeEnabled.isChecked)
        }

        binding.title.afterTextChanged { timerRecordingViewModel.recording.title = it }
        binding.name.afterTextChanged { timerRecordingViewModel.recording.name = it }
        binding.directory.afterTextChanged { timerRecordingViewModel.recording.directory = it }
        binding.isEnabled.setOnCheckedChangeListener { _, isChecked ->
            timerRecordingViewModel.recording.isEnabled = isChecked
        }
    }

    private fun handleTimeEnabledClick(checked: Boolean) {
        Timber.d("Setting time enabled ${binding.timeEnabled.isChecked}")
        timerRecordingViewModel.isTimeEnabled = checked

        binding.startTimeLabel.visibleOrGone(checked)
        binding.startTime.visibleOrGone(checked)
        binding.startTime.isEnabled = checked

        binding.stopTimeLabel.visibleOrGone(checked)
        binding.stopTime.visibleOrGone(checked)
        binding.stopTime.isEnabled = checked
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        menu.forEach { it.isVisible = false }
        menu.findItem(R.id.menu_save)?.isVisible = true
        menu.findItem(R.id.menu_cancel)?.isVisible = true
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.save_cancel_options_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                cancel()
                true
            }
            R.id.menu_save -> {
                save()
                true
            }
            R.id.menu_cancel -> {
                cancel()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun save() {
        if (timerRecordingViewModel.recording.title.isNullOrEmpty()) {
            context?.sendSnackbarMessage(R.string.error_empty_title)
            return
        }

        val intent = timerRecordingViewModel.getIntentData(requireContext(), timerRecordingViewModel.recording)

        // Add the recording profile if available and enabled
        if (profile != null && htspVersion >= 16 && binding.dvrConfig.text.isNotEmpty()) {
            intent.putExtra("configName", binding.dvrConfig.text.toString())
        }

        // Update the recording in case the id is not empty, otherwise add a new one.
        // When adding a new recording, the id is an empty string as a default.
        if (timerRecordingViewModel.recording.id.isNotEmpty()) {
            intent.action = "updateTimerecEntry"
            intent.putExtra("id", timerRecordingViewModel.recording.id)
        } else {
            intent.action = "addTimerecEntry"
        }
        activity?.startService(intent)
        //activity?.finish()
        activity?.supportFragmentManager?.popBackStack()
    }

    private fun cancel() {
        Timber.d("cancel")
        // Show confirmation dialog to cancel
        context?.let {
            MaterialDialog(it).show {
                message(R.string.cancel_add_recording)
                positiveButton(R.string.discard) {
                    //activity?.finish()
                    Timber.d("discarding popping back stack")
                    activity?.supportFragmentManager?.popBackStack()
                }
                negativeButton(R.string.cancel) {
                    Timber.d("dismissing dialog")
                    dismiss()
                }
            }
        }
    }

    override fun onChannelSelected(channel: Channel) {
        timerRecordingViewModel.recording.channelId = channel.id
        binding.channelName.text = channel.name
    }

    override fun onPrioritySelected(which: Int) {
        timerRecordingViewModel.recording.priority = which
        context?.let {
            binding.priority.text = getPriorityName(it, timerRecordingViewModel.recording.priority)
        }
    }

    override fun onProfileSelected(which: Int) {
        binding.dvrConfig.text = recordingProfilesList[which]
        timerRecordingViewModel.recordingProfileNameId = which
    }

    override fun onTimeSelected(milliSeconds: Long, tag: String?) {
        if (tag == "startTime") {
            timerRecordingViewModel.startTimeInMillis = milliSeconds
            // If the start time is after the stop time, update the stop time with the start value
            if (milliSeconds > timerRecordingViewModel.stopTimeInMillis) {
                timerRecordingViewModel.stopTimeInMillis = milliSeconds
            }
        } else if (tag == "stopTime") {
            timerRecordingViewModel.stopTimeInMillis = milliSeconds
            // If the stop time is before the start time, update the start time with the stop value
            if (milliSeconds < timerRecordingViewModel.recording.startTimeInMillis) {
                timerRecordingViewModel.startTimeInMillis = milliSeconds
            }
        }

        binding.startTime.text = getTimeStringFromTimeInMillis(timerRecordingViewModel.startTimeInMillis)
        binding.stopTime.text = getTimeStringFromTimeInMillis(timerRecordingViewModel.stopTimeInMillis)
    }

    override fun onDateSelected(milliSeconds: Long, tag: String?) {
        // NOP
    }

    override fun onDaysSelected(selectedDays: Int) {
        timerRecordingViewModel.recording.daysOfWeek = selectedDays
        context?.let {
            binding.daysOfWeek.text = getSelectedDaysOfWeekText(it, selectedDays)
        }
    }

    override fun onBackPressed() {
        Timber.d("onBackPressed")
        cancel()
    }

    companion object {
        fun newInstance(id: String = ""): TimerRecordingAddEditFragment {
            val f = TimerRecordingAddEditFragment()
            if (id.isNotEmpty()) {
                f.arguments = Bundle().also { it.putString("id", id) }
            }
            return f
        }
    }
}
